iceoryx_examples/icedelivery/iox_subscriber_typed_modern.cpp

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2019 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_posh/popo/modern_api/typed_subscriber.hpp"
+#include "iceoryx_posh/popo/modern_api/untyped_subscriber.hpp"
+#include "iceoryx_posh/popo/guard_condition.hpp"
+#include "iceoryx_posh/popo/wait_set.hpp"
+#include "iceoryx_posh/runtime/posh_runtime.hpp"
+#include "topic_data.hpp"
+
+#include <chrono>
+#include <csignal>
+#include <iostream>
+
+bool killswitch = false;
+iox::popo::GuardCondition shutdownGuard;
+
+static void sigHandler(int f_sig [[gnu::unused]])
+{
+    // caught SIGINT, now exit gracefully
+    killswitch = true;
+    shutdownGuard.trigger(); // unblock waitsets
+}
+
+void subscriberHandler(iox::popo::WaitSet& waitSet)
+{
+    // run until interrupted
+    while(!killswitch)
+    {
+        auto triggeredConditions = waitSet.wait();
+        for(auto& condition : triggeredConditions)
+        {
+            auto subscriber = dynamic_cast<iox::popo::TypedSubscriber<Position>*>(condition);
```

Comment:
`static_cast` should be enough? We don't have RTTI available, so a dynamic_cast doesn't make sense here.

Reply:
Can the condition also be the `shutdownGuard`? I guess no? I have to get up to date with the new API :)

Reply:
It can, which is why I did the dynamic_cast.
The point of the cast is to check the type of condition. I can't use a static cast.
I'm not too happy with this approach though, anyone got an alternative ?

Reply:
Yeah, it can be a `shutdownGuard` / `GuardCondition` IIRC. Should we introduce a flag in the base class for this? Or fast forwarding, if we have additional new Conditions, how do we differentiate here?

Reply:
One way would be to use unique ids.
The waitset could return a list of unique IDs then the user can determine which conditions are triggered by matching them.

Reply:
Yes, I wanted to propose something like that
```
#include <iostream>

class Base {
public:
    uint32_t getConditionId() {
        return m_conditionId;
    }
    
    template <typename T>
    static uint32_t getConditionIdForType() {
        static uint32_t conditionId = conditionIdGenerator();
        return conditionId;
    }
    
protected:
    Base (uint32_t conditionId) : m_conditionId(conditionId) {}
        
private:
    static uint32_t conditionIdGenerator() {
        static uint32_t conditionId = 0; 
        conditionId++;
        return conditionId;
    }
    
    uint32_t m_conditionId {0};
};

class Foo : public Base {
public:
    Foo() : Base(Base::getConditionIdForType<Foo>()) {}
};

class Bar : public Base {
public:
    Bar() : Base(Base::getConditionIdForType<Bar>()) {}
};

int main() {
    Foo foo;
    
    Base* base = &foo;
    
    if (base->getConditionId() == Base::getConditionIdForType<Foo>()) {
        std::cout << "It's a Foo" << std::endl;
    }
    
    if (base->getConditionId() != Base::getConditionIdForType<Bar>()) {
        std::cout << "It's not a Bar" << std::endl;
    }
    
    return EXIT_SUCCESS;
}
```
I couldn't resist and write some code. Something like that might be feasible. In the end we reinvent RTTI. For the TypeId we could use the newtype pattern.

Reply:
How about this approach:

Reply:
This shall be addressed in #341